### PR TITLE
DATAUP-725: Return Optionals vs. nulls from ObjectIdentifier API

### PR DESCRIPTION
**Pull request by MrCreosote**
Avoid the billion dollar mistake

### CODE REVIEW

File: src/us/kbase/workspace/database/ObjectIdentifier.java [hunk 1]
```diff
@@ -17,7 +18,6 @@
 	// TODO TEST unittests (move from WorkspaceTest, mostly covered there)
 	// also need to test toString code, but not super high priority
 	// TODO NOW JAVADOC
-	// TODO CODE return Optionals instead of nulls. That's a big change...
```
Comment (MrCreosote) [inline src/us/kbase/workspace/database/ObjectIdentifier.java:20]: Eh, wasn't that big when I looked into it

File: src/us/kbase/workspace/database/ObjectIdentifier.java [hunk 2]
```diff
@@ -411,9 +411,9 @@ private Builder(final WorkspaceIdentifier wsi) {
 
 		private Builder(final ObjectIdentifier oi) {
 			this.wsi = requireNonNull(oi, "oi").getWorkspaceIdentifier();
-			this.id = oi.getID() == null ? -1 : oi.getID();
-			this.name = oi.getName();
-			this.version = oi.getVersion() == null ? -1 : oi.getVersion();
+			this.id = oi.getID().orElse(-1L);
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/ObjectIdentifier.java:414]: `orElse` -- fantastic!

**Review by ialarmedalien — APPROVED** [review]
Looks fine

File: src/us/kbase/workspace/database/Reference.java [hunk 3]
```diff
@@ -43,8 +43,8 @@ public Reference(final String ref) {
 					"ref %s is not an absolute reference", ref));
 		}
 		workspaceID = oi.getWorkspaceIdentifier().getId();
-		objectID = oi.getID();
-		version = oi.getVersion();
+		objectID = oi.getID().get();
+		version = oi.getVersion().get();
```
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/us/kbase/workspace/database/Reference.java:47]: Yeah, in some cases Optionals are great (see `orElse` above) and in some cases they add extra steps. Overall, they're a lot better than `null`s though.

**Comment by MrCreosote** [comment]
@ialarmedalien  I had to push some changes to get the java 8 tests to pass. Can you re-review?

**Comment by MrCreosote** [comment]
Since emoji reactions can be removed and @ialarmedalien is known for her devious and generally reprobate nature:
image

**Pull request merged**